### PR TITLE
Using JInput to get server information

### PR DESCRIPTION
**Pull request by wilsonge**
This ports back some fixes from the framework to grab data from JInput rather than $_SERVER which adds an additional layer of sanity checks to the data.

There's no easy way to test this. Just browse around and ensure that any thing URL and session related continue to work as usual.

### CODE REVIEW

File: libraries/joomla/application/web.php [hunk 1]
```diff
@@ -840,8 +840,10 @@ protected function checkHeadersSent()
 	 */
 	protected function detectRequestUri()
 	{
+		$serverSSLVar = $this->input->server->getString('HTTPS', '');
+
 		// First we need to detect the URI scheme.
-		if (isset($_SERVER['HTTPS']) && !empty($_SERVER['HTTPS']) && (strtolower($_SERVER['HTTPS']) != 'off'))
+		if (isset($serverSSLVar) && !empty($serverSSLVar) && (strtolower($serverSSLVar) != 'off'))
```
Comment (joomdonation) [inline libraries/joomla/application/web.php]: I think isset($serverSSLVar) is not needed here.

Comment (Bakual) [inline libraries/joomla/application/web.php]: @joomdonation is right for two reasons:
* `$serverSSLVar` is always set just a few lines above, so `isset()` will never fail
* `empty()` already has an implied `isset()` integrated. So we're checking it twice.

File: libraries/joomla/application/web.php [hunk 2]
```diff
@@ -856,22 +858,26 @@ protected function detectRequestUri()
...
 		else
 		{
 			// IIS uses the SCRIPT_NAME variable instead of a REQUEST_URI variable... thanks, MS
-			$uri = $scheme . $_SERVER['HTTP_HOST'] . $_SERVER['SCRIPT_NAME'];
+			$uri = $scheme . $this->input->server->getString('HTTP_HOST') . $this->input->server->getString('SCRIPT_NAME');
+			$queryString = $this->input->server->getString('QUERY_STRING', '');
 
 			// If the QUERY_STRING variable exists append it to the URI string.
-			if (isset($_SERVER['QUERY_STRING']) && !empty($_SERVER['QUERY_STRING']))
+			if (isset($queryString) && !empty($queryString))
```
Comment (Bakual) [inline libraries/joomla/application/web.php]: Same as above applies here. `isset()` isn't needed as `$queryString` is always set and `empty()` checks it as well.

File: libraries/joomla/uri/uri.php [hunk 3]
```diff
@@ -62,8 +62,11 @@ public static function getInstance($uri = 'SERVER')
 			// Are we obtaining the URI from the server?
 			if ($uri == 'SERVER')
 			{
+				$input = new JInput;
+				$serverSSLVar = $input->server->getString('HTTPS', '');
+
 				// Determine if the request was over SSL (HTTPS).
-				if (isset($_SERVER['HTTPS']) && !empty($_SERVER['HTTPS']) && (strtolower($_SERVER['HTTPS']) != 'off'))
+				if (isset($serverSSLVar) && !empty($serverSSLVar) && (strtolower($serverSSLVar) != 'off'))
```
Comment (Bakual) [inline libraries/joomla/uri/uri.php]: Same here, `isset()` isn't needed.

**Comment by brianteeman** [comment]
Something not quite right here George as seen by the unit test results<hr /><sub>This comment was created with the <a href="https://github.com/joomla/jissues">J!Tracker Application</a> at <a href="https://issues.joomla.org/tracker/joomla-cms/8680">issues.joomla.org/joomla-cms/8680</a>.</sub>

**Comment by wilsonge** [comment]
This needs to be done - and I know what the problem is - but I need to think about the "right" solution - will come back to it another day

**Pull request closed without merging**